Control batch sizes during execute

dmitchell opened the pull request:
@arsduo please accept this revision which limits the batch sizes to FB's limits

dmitchell commented:
I'm going to assume the failure for JRuby19 loading rake is completely separate from this.

dmitchell commented:
Another thing this PR fixes is recursive calls. The current master resubmits the whole batch set plus the additions if you do something like the following, but this PR removes each submitted call from `batch_calls` and thus avoids the concatenation problem:

```
  api.batch do |batch_api|
     MyObject.each { |obj|
        batch_api.get_object(obj.external_id, {fields: '..'}) { |ret|
           batch_api.put_connections(obj.external_id, '', {foo: new_value_from(ret)}) { .. }
        }
     }
  end 
```

dmitchell commented:
Re the build failures on faraday going into infinite recursion, I had that problem until I reinstalled the faraday gem. idk if something changed in the gem. The recursion makes no sense from the source code.

arsduo commented:
Hi @dmitchell -- thanks for the awesome pull request! Sorry about the delay in getting back to you and sorry about the some issues with the testing. I've fixed the test failures (and removed a bunch of now-dead code), which should resolve that. (Looks like you unfortunately have to fix some conflicts, sorry 😐.)

I've looked over the PR and will give it another review soon. Would you be willing to split this into two separate pull requests? It's really two different changes, one more involved than the other; I'd be happy to get the batch size limit change in quickly while reviewing the recursive changes.

dmitchell commented:
Yes I'll split it. Thanks for the base fixes.

On Thu, Feb 16, 2017 at 7:51 PM Alex Koppel <notifications@github.com>
wrote:

> Hi @dmitchell <https://github.com/dmitchell> -- thanks for the awesome
> pull request! Sorry about the delay in getting back to you and sorry about
> the some issues with the testing. I've fixed the test failures (and removed
> a bunch of now-dead code), which should resolve that. (Looks like you
> unfortunately have to fix some conflicts, sorry 😐.)
>
> I've looked over the PR and will give it another review soon. Would you be
> willing to split this into two separate pull requests? It's really two
> different changes, one more involved than the other; I'd be happy to get
> the batch size limit change in quickly while reviewing the recursive
> changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/arsduo/koala/pull/555#issuecomment-280515015>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAdDD3d6Ny7wjZ4jwj82NjLePo-K5W0Pks5rdO74gaJpZM4Ks-7q>
> .
>

dmitchell commented on lib/koala/api/graph_batch_api.rb:
```diff
@@ -8,6 +8,11 @@ class GraphBatchAPI < API
       include GraphAPIMethods
 
+      # Limits from @see https://developers.facebook.com/docs/marketing-api/batch-requests/v2.8
+      MAX_AD_SUBMISSIONS = 10
```
From what I can tell, altho FB is hard about `MAX_CALLS`, `MAX_AD_SUBMISSIONS` may be just a recommendation.

dmitchell commented:
Going to create 2 new PRs with the rebase and kill this one
